### ISSUE TITLE
SVD on GPU is slower than SVD on CPU

### ISSUE THREAD

**Issue by hzhangxyz** [tensorflow/tensorflow#13603]
OS:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: CentOS release 7.4.1708
- **TensorFlow installed from (source or binary)**: From source
- **Python version**: 2.7.13
- **Bazel version**: 0.6.1
- **CUDA/cuDNN version**: CUDA 8.0/cuDNN 6.0.21
- **GPU model and memory**: GeForce GTX 950M, memory 4GB

output of `tf_env_collect.sh`
```

== cat /etc/issue ===============================================
Linux zhanghao 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="7 (Core)"
VERSION_ID="7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

== are we in docker =============================================
No

== compiler =====================================================
c++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
Copyright © 2015 Free Software Foundation, Inc.
本程序是自由软件；请参看源代码的版权声明。本软件没有任何担保；
包括没有适销性和某一专用目的下的适用性担保。

== uname -a =====================================================
Linux zhanghao 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================

== check for virtualenv =========================================
False

== tensorflow import ============================================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named tensorflow

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Tue Oct 10 16:36:08 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.90                 Driver Version: 384.90                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 950M    Off  | 00000000:0A:00.0 Off |                  N/A |
| N/A   45C    P0    N/A /  N/A |      0MiB /  4044MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================

== cat /etc/issue ===============================================
Linux zhanghao 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="7 (Core)"
VERSION_ID="7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

== are we in docker =============================================
No

== compiler =====================================================
c++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
Copyright © 2015 Free Software Foundation, Inc.
本程序是自由软件；请参看源代码的版权声明。本软件没有任何担保；
包括没有适销性和某一专用目的下的适用性担保。

== uname -a =====================================================
Linux zhanghao 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.12.1)
protobuf (3.4.0)
tensorflow (1.4.0rc0)
tensorflow-tensorboard (0.4.0rc1)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.4.0-rc0
tf.GIT_VERSION = v1.3.0-rc1-3111-g4196d6d
tf.COMPILER_VERSION = v1.3.0-rc1-3111-g4196d6d
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH /usr/local/cuda/lib64/:/usr/local/cuda/lib64/stubs/:/usr/local/cuda/extras/CUPTI/lib64/:/usr/local/cuda/nvvm/lib64/:/usr/lib64/nvidia/:/opt/intel/compilers_and_libraries_2017.4.196/linux/compiler/lib/intel64:/opt/intel/compilers_and_libraries_2017.4.196/linux/compiler/lib/intel64_lin:/opt/intel/compilers_and_libraries_2017.4.196/linux/mpi/intel64/lib:/opt/intel/compilers_and_libraries_2017.4.196/linux/mpi/mic/lib:/opt/intel/compilers_and_libraries_2017.4.196/linux/ipp/lib/intel64:/opt/intel/compilers_and_libraries_2017.4.196/linux/compiler/lib/intel64_lin:/opt/intel/compilers_and_libraries_2017.4.196/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2017.4.196/linux/tbb/lib/intel64/gcc4.7:/opt/intel/debugger_2017/iga/lib:/opt/intel/debugger_2017/libipt/intel64/lib:/opt/intel/compilers_and_libraries_2017.4.196/linux/daal/lib/intel64_lin:/opt/intel/compilers_and_libraries_2017.4.196/linux/daal/../tbb/lib/intel64_lin/gcc4.4
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Tue Oct 10 16:36:37 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.90                 Driver Version: 384.90                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 950M    Off  | 00000000:0A:00.0 Off |                  N/A |
| N/A   45C    P0    N/A /  N/A |      0MiB /  4044MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart.so.8.0.61
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart_static.a
```

output of `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
```
('v1.3.0-rc1-3111-g4196d6d', '1.4.0-rc0')
```

### Describe the problem

SVD on GPU is slower than SVD on CPU

### Source code / logs

file main.py
```
import tensorflow as tf
import numpy as np
import sys

D = 1024
dA = np.random.normal(size=(D,D))

dev = "/gpu:0" if len(sys.argv)==1 else "/cpu:0"

with tf.device(dev):
    A = tf.placeholder(shape=(D,D),dtype=tf.float32)
    S, U, V = tf.svd(A)

config = tf.ConfigProto()
config.log_device_placement = True
config.graph_options.optimizer_options.global_jit_level=tf.OptimizerOptions.ON_1
sess = tf.Session(config=config)

for _ in xrange(10):
    dS, dU, dV = sess.run((S, U, V), feed_dict={A:dA})
```

## run on GPU
`time python main.py`
```
2017-10-10 16:28:49.047703: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-10-10 16:28:49.048176: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties:
name: GeForce GTX 950M major: 5 minor: 0 memoryClockRate(GHz): 1.124
pciBusID: 0000:0a:00.0
totalMemory: 3.95GiB freeMemory: 3.91GiB
2017-10-10 16:28:49.048205: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0)
Device mapping:
/job:localhost/replica:0/task:0/device:XLA_GPU:0 -> device: XLA_GPU device
/job:localhost/replica:0/task:0/device:XLA_CPU:0 -> device: XLA_CPU device
/job:localhost/replica:0/task:0/device:GPU:0 -> device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0
2017-10-10 16:28:49.064960: I tensorflow/core/common_runtime/direct_session.cc:299] Device mapping:
/job:localhost/replica:0/task:0/device:XLA_GPU:0 -> device: XLA_GPU device
/job:localhost/replica:0/task:0/device:XLA_CPU:0 -> device: XLA_CPU device
/job:localhost/replica:0/task:0/device:GPU:0 -> device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0

Svd: (Svd): /job:localhost/replica:0/task:0/device:GPU:0
2017-10-10 16:28:49.067234: I tensorflow/core/common_runtime/placer.cc:874] Svd: (Svd)/job:localhost/replica:0/task:0/device:GPU:0
Placeholder: (Placeholder): /job:localhost/replica:0/task:0/device:GPU:0
2017-10-10 16:28:49.067302: I tensorflow/core/common_runtime/placer.cc:874] Placeholder: (Placeholder)/job:localhost/replica:0/task:0/device:GPU:0
2017-10-10 16:28:49.074053: I tensorflow/core/kernels/cuda_solvers.cc:159] Creating CudaSolver handles for stream 0x488e860
python main.py  27.50s user 2.30s system 100% cpu 29.658 total
```

## run on CPU
`time python main.py -`
```
2017-10-10 16:29:53.252138: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-10-10 16:29:53.252572: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties:
name: GeForce GTX 950M major: 5 minor: 0 memoryClockRate(GHz): 1.124
pciBusID: 0000:0a:00.0
totalMemory: 3.95GiB freeMemory: 3.91GiB
2017-10-10 16:29:53.252600: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0)
Device mapping:
/job:localhost/replica:0/task:0/device:XLA_GPU:0 -> device: XLA_GPU device
/job:localhost/replica:0/task:0/device:XLA_CPU:0 -> device: XLA_CPU device
/job:localhost/replica:0/task:0/device:GPU:0 -> device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0
2017-10-10 16:29:53.269242: I tensorflow/core/common_runtime/direct_session.cc:299] Device mapping:
/job:localhost/replica:0/task:0/device:XLA_GPU:0 -> device: XLA_GPU device
/job:localhost/replica:0/task:0/device:XLA_CPU:0 -> device: XLA_CPU device
/job:localhost/replica:0/task:0/device:GPU:0 -> device: 0, name: GeForce GTX 950M, pci bus id: 0000:0a:00.0, compute capability: 5.0

Svd: (Svd): /job:localhost/replica:0/task:0/device:CPU:0
2017-10-10 16:29:53.271505: I tensorflow/core/common_runtime/placer.cc:874] Svd: (Svd)/job:localhost/replica:0/task:0/device:CPU:0
Placeholder: (Placeholder): /job:localhost/replica:0/task:0/device:CPU:0
2017-10-10 16:29:53.271544: I tensorflow/core/common_runtime/placer.cc:874] Placeholder: (Placeholder)/job:localhost/replica:0/task:0/device:CPU:0
python main.py -  34.33s user 10.68s system 621% cpu 7.241 total
```

**Comment by hzhangxyz** [comment]
It seems during some time of GPU SVD, there is frequent data moving between Host and Device. I don't know whether it is the reason why GPU SVD is slow.

**Comment by shamanDevel** [comment]
Each computation of the SVD on the GPU requires one memcpy from Device to the Host: the status report if the algorithm was successful needs to be copied from the GPU to the CPU to be checked.
This can in principle be a bottleneck, but can't be avoided.

But for me, the logs indicate that the GPU version is indeed faster than the CPU if you compare the timings. See my comment in the pull request for more details.

**Comment by hzhangxyz** [comment]
As another clear sample:
```
import tensorflow as tf
import numpy as np
from tensorflow.python.client import timeline

options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
sess = tf.Session()

D = 1024
dA = np.random.normal(size=(D,D))
A = tf.placeholder(shape=(D,D),dtype=tf.float32)

with tf.device("/cpu:0"):
    C = tf.svd(A)

with tf.device("/gpu:0"):
    G = tf.svd(A)

_ = sess.run(C+G, feed_dict={A:dA}, options=options, run_metadata=run_metadata)

fetched_timeline = timeline.Timeline(run_metadata.step_stats)
chrome_trace = fetched_timeline.generate_chrome_trace_format()
with open('timeline.json', 'w') as f:
    f.write(chrome_trace)
```

run it and get `timeline.json`, and open it via chrome then get it:

![image](https://user-images.githubusercontent.com/11623447/31384419-8d85faae-ad84-11e7-91a2-7311bd228db8.png)

those memcpy are obviously called by SVD. there is nothing else in the program.

Then zoom in, we can find there is many small memcpy here

![image](https://user-images.githubusercontent.com/11623447/31384472-c5672c68-ad84-11e7-943f-e471cd2cf1ca.png)

**Comment by yaroslavvb** [comment]
@hzhangxyz your benchmark is confusing because you have both CPU and GPU SVD in a single session run. It's better to isolate to a single op if the goal is to show that op kernel is slow.

I have a benchmark in https://github.com/tensorflow/tensorflow/issues/13222#issuecomment-331642490 which isolates to just GPU SVD and rules out memory transfers.

In that example (n=1534, float32), TF CPU runs about 4.6 slower than corresponding version MKL-enabled numpy, TF GPU version runs about  21x slower.

in commit: https://github.com/tensorflow/tensorflow/commit/22a886b

**Comment by yaroslavvb** [comment]
BTW, I updated [benchmark](https://github.com/yaroslavvb/stuff/blob/master/gpu_svd_bench.py) with PyTorch numbers. PyTorch and TF get similar performance on CPU, but very different on GPU (PyTorch ran 12x faster)

**Comment by yaroslavvb** [comment]
I've updated [benchmark](https://github.com/yaroslavvb/stuff/blob/master/gpu_svd_bench.py) to compare against PyTorch svd as well. CPU performance is similar between PyTorch and TF, but GPU performance was 12x faster in PyTorch

**Comment by reedwm** [comment]
@rmlarsen can you or someone look into why PyTorch is faster, and why the GPU op is slower than the CPU op?

**Comment by shamanDevel** [comment]
This really intrigues me now. I'll also have a look at PyTorch to see what algorithm they use.

**Comment by rmlarsen** [comment]
This is strange. Let me take a look.

**Comment by shamanDevel** [comment]
Wow, this is a good point. According to this benchmark http://www.netlib.org/lapack/lug/node71.html, GESDD is way faster than GESVD. Sadly, GESDD is not implemented in cuSolver, so we would have to use Magma instead. 
It appears that Magma is also currently under more active development (If I just look at the publications), and they have some other highly interesting algorithm implementations. Maybe replace cuSolver completely by Magma? Or add a compilation switch on which implementation should be used? The details would then need to be hidden in the cuSolver wrapper (which probably should be renamed then).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by hzhangxyz** [comment]
it seems all kind of decomposition is not proper for gpu?
as for qr, cpu is also faster than gpu

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shamanDevel** [comment]
Hello,
I'm back from a long vacation and I'd like to add my piece to the discussion:

- Yes, matrix decompositions are very often slower on the GPU than on the CPU. These are simply problems that are hard to parallelize on the GPU architecture. 
- Yes, Eigen without MKL (that's what TF uses on the CPU) is slower than numpy with MKL

I would vote for keeping the GPU version of the SVD, because for some CPU-GPU configurations and problem sizes, it is indeed faster. That was the case for me when I proposed the Pull Request.
Is it possible to declare the CPU version as the default one and only use the GPU version if explicitly requested? If I'm not wrong, TF uses always the GPU version if available and if nothing else is specified.

**Comment by PGadoury** [comment]
Hi,
I'm writing some code that needs to generate on the order of millions of small SVD decompositions. I believe since I updated to 1.6 tensorflow is now defaulting to CPU to calculate these? I'd love to be able to benchmark or verify, but I'm getting a "no kernel found" error if I try to hard-assign my GPU on the calculation.

It seems most of you are able to switch freely between the two, so I am genuinely perplexed. What complicates matters is I have recently upgraded to Windows 10 and updated my Nvidia drivers, so that might also be the case.

Sorry to bother with what amounts to a tech support question :/

Cheers!

**Comment by hzhangxyz** [comment]
@PGadoury , If you want specify device for tensorflow, there is tutorial [here](https://www.tensorflow.org/programmers_guide/using_gpu), just add `with tf.device('/cpu:0'):` and so on. but first of all, you need to ensure your tensorflow compiled with gpu support.

**Comment by rmlarsen** [comment]
NVIDIA has worked on (and are continuing to do so) improving the performance of their symmetric eigensolvers and SVD in CUDA 9, especially 9.1, which also adds batched interfaces. We will switch to using the faster versions and/or the batched interfaces as those versions of CUDA become supported by TensorFlow.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 95 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 110 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 125 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 140 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shamanDevel** [comment]
As mentioned over and over, SVD and Eigenvalue Decomposition is very tricky to efficiently parallelize on the GPU. As rmlarsen mentioned above, NVIDIA continues the development of cuSolver, so we can hope for some performance improvements in the future. For now, the CPU version might be faster in most cases.
As always: Profiling! If the CPU version is faster than the GPU version, then use the CPU version.

**Comment by Bihaqo** [comment]
It can be painful to wrap each of the many occurrences of tf.svd and tf.qr with `with tf.devices(":cpu")`. Is there a nice way to do a program-wide flag for moving all SVD ops onto CPU?

**Comment by shamanDevel** [comment]
That would be an awesome feature. Some per-operation flag that specifies whether the GPU or CPU version should be selected by default. To my knowledge, if a GPU version is available, that one is always selected. @rmlarsen what do you think?

**Comment by brianwa84** [comment]
You can already do this:
https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/training/device_setter.py#L97
https://stackoverflow.com/questions/39666845/how-does-tf-train-replica-device-setter-work

Brian Patton | Software Engineer | bjp@google.com | +1 (360) 903-9003

**Comment by yaroslavvb** [comment]
GPU version is so slow right now I don't think it makes sense for anyone. Here are some numbers from a recent benchmark -- https://github.com/yaroslavvb/stuff/tree/master/linalg-benchmark

**Comment by PGadoury** [comment]
Your benchmark assumes a single large SVD operation as a use case. TF might well be more efficient to calculate a multitude of smaller SVD's in parallel...

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 77 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by AlexeyG** [comment]
Has there been any progress on this? I am now using SVD for log-det computation (it's seems to be more numerically stable than other slog_det or det), but it's also considerably slower than ways of computing the log-det.

Switching to CPU and using a custom gradient helps (~ 15x speed-up), but would be great if it can be even faster.

**Comment by jbloom22** [comment]
We've recast SVD as the problem of minimizing an L_2-regularized linear autoencoder, making it far more amenable to hardware and software optimizations in TF.  With SGD, the approach is far more similar to randomized SVD rather than GESDD. Would love to see some experiments with the full array of dirty deep learning tricks!

https://github.com/danielkunin/Regularized-Linear-Autoencoders
https://twitter.com/jbloom22/status/1089675387163021313

**Comment by smit-hinsu** [comment]
Using the example from #28518, it seems that there is further regression in TensorFlow version 1.13 compared to 1.12. I don't see any significant change in the GPU kernel for the SVD op so this is likely caused by upgrade to CUDA 10.0 in TF version 1.13 from 9.0.

$ time docker run --runtime=nvidia -v ~/svd:/svd/ tensorflow/tensorflow:1.12.0-gpu python /svd/graph_v1.py
real	0m27.925s
user	0m0.036s
sys	0m0.040s

$ time docker run --runtime=nvidia -v ~/svd:/svd/ tensorflow/tensorflow:1.13.1-gpu python /svd/graph_v1.py
real	1m4.272s
user	0m0.032s
sys	0m0.044s

**Comment by sleighsoft** [comment]
Did you also benchmark against 2.0a?

**Comment by seanslice** [comment]
Hello, I am having a different problem in that eigen decomposition is extremely slow in TF for both CPU and GPU. Here is a CPU example for a moderate sized matrix (400 x 400):
```
import numpy as np
import scipy as sp
import tensorflow as tf
from time import time

A = np.random.randn(400, 400)
A_tf = tf.constant(A)

cur = time()
d, v = sp.linalg.eig(A)
print(f'sp: {time() - cur:4.2f} s')

cur = time()
d, v = np.linalg.eig(A)
print(f'np: {time() - cur:4.2f} s')

cur = time()
d, v = tf.linalg.eig(A_tf)
print(f'tf: {time() - cur:4.2f} s')
```
With output:
```
sp: 0.09 s
np: 0.08 s
tf: 5.04 s
```
Any ideas?

**Comment by rmothukuru** [comment]
@hzhangxyz,
This performance issue seems to be resolved in the latest version of **`Tensorflow, 2.5`** because of subsequent changes and upgrades in the **`CUDA`** and **`CuDNN`** versions.

We have executed your code in [Google Colab with GPU Runtime](https://colab.research.google.com/gist/rmothukuru/6f92a74a1a97d7604123f075e9c8189b/gh_13603.ipynb) and observed that there is an **`improvement in performance`**. Can you please validate our observation so that we can close this issue? Thanks!

**Comment by Randl** [comment]
I'd note that TF is still significantly slower than, for example, numpy (even without MKL), but at least GPU is faster than CPU now
```Python version:  3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33) 
[GCC 9.3.0]
CPU version:  model name	: Intel(R) Xeon(R) Silver 4208 CPU @ 2.10GHz
CPU logical cores: 32
CPU physical cores: 16
CPU physical sockets:  2
MKL version b'Intel(R) oneAPI Math Kernel Library Version 2021.2-Product Build 20210312 for Intel(R) 64 architecture applications'
Scipy version:  1.6.3
Numpy version:  1.19.5
TF version:  v2.5.0-rc3-213-ga4dfb8d1a71 https://github.com/tensorflow/tensorflow/commit/a4dfb8d1a71
Timing in ms for 1534 x 1534 SVD
numpy default        min:   447.89, median:   453.21, mean:   469.07
numpy gesvd          min:  1930.24, median:  1934.58, mean:  1935.91
numpy gesdd          min:   449.35, median:   450.08, mean:   450.22
skipping torch
TF CPU               min:  1702.51, median:  1707.78, mean:  1709.30
TF GPU               min:  1068.35, median:  1069.48, mean:  1145.60```

**Comment by Bihaqo** [comment]
Thank you, I think this indeed resolves the problem!

Evgeniy is right, it's still not perfect, but probably worth tracking in a
separate issue.

**Comment by hzhangxyz** [comment]
@rmothukuru  At least GPU is not slower than CPU for svd. Being slower than numpy(as @Randl said, I did not test) is another problem.
